### DISCUSSION TITLE
How to create Multi-stage Docker image with Kafka-connect? Critical Issues Strimzi Kafka Docker Image

### DISCUSSION THREAD

**Discussion by imrohankataria** [strimzi/strimzi-kafka-operator#8939] [Q&A]
Hi,

My Team is planned for Go-live next week but the latest strimzi has issues in the image, and thats coming as a blocker for our release.
Any suggestions on how to fix these issues? I also looked at https://quay.io/repository/strimzi/kafka/manifest/sha256:2fed0b67104b52214cde95cd6727a5609c728d63e601e19e32e7bfe95186abc3?tab=vulnerabilities How are these issues going to be addressed? and when? Thanks

Image scan by security team - 
<img width="1390" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/e547ac52-6c46-427e-8cfd-40d52cd4ff19">

**UPDATE:**

My Team is using Strimzi Kafka connect in enterprise setting, and because of the image straight from quay.io we are seeing lots of vulnerabilities in the image.

To resolve those CVEs, I have been able to drop the folders/utilities that are not being used eg. kafa-exporter, cruise-control, etc

But the issues are still coming up because of Shasum history in the docker image. One resolution I tried before was squashing the images, but it is not working anymore. 

Another approach is creating a Multi-stage image, which Im failing to create. 

Will appreciate your help on this.

Code I have been trying to use -

```
FROM quay.io/strimzi/kafka:0.36.1-kafka-3.5.1 as builder

# Switch to root user to make modifications
USER root

# Remove unwanted components
RUN rm -rf /opt/kafka-exporter /opt/cruise-control

# Remove the old guava library
RUN rm /opt/kafka/libs/guava-30.1-jre.jar

# # Stage 2: Final Alpine-based image
FROM alpine:latest

# # Create Kafka user with similar properties to Strimzi's Kafka user
RUN addgroup -g 1001 -S kafka && adduser -u 1001 -S kafka -G kafka

# # Copy necessary files from the builder stage, retaining permissions
COPY --from=builder --chown=kafka:kafka /opt/kafka /opt/kafka

# # Add full permissions for the 'kafka' group to the /opt/kafka directory
RUN chmod -R g+rwx /opt/kafka

# Switch to Kafka user
USER kafka
```

**Comment by imrohankataria** [comment]
@scholzj Hello Sir, any suggestions?

**Reply by scholzj** [reply]
They need to be addressed at their source. Your screenshot does not share where they were found, so it is not obvious. The SnakeYAML is likely coming from Cruise Control. Not sure where the Go issue comes from, you would need to share in what files was it identified. Possibly in the Kafka Exporter?

**Reply by imrohankataria** [reply]
<img width="545" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/37dc1b1a-fe23-414b-b2bd-dc8e5ac3ee65">

**Reply by imrohankataria** [reply]
<img width="478" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/4261f508-b313-4a51-8379-68a03ddeb190">

**Reply by imrohankataria** [reply]
Golang is from Kafka Exporter, We don't use Kafka exporter anyway. Is it possible to take it out of image?

**Reply by scholzj** [reply]
If you don't use it, you can rebuild the image to remove it, yes. Either from the sources, or maybe easier, just write your own Dockerfile which starts from the Strimzi image and deletes it.

**Reply by imrohankataria** [reply]
deleting from docker is going to be an easier solution, can how to go about the other vulnerabilities? can the be updated in docker?

**Reply by scholzj** [reply]
If you don't use Cruise Control, you can delete it as well. If you use it, I think you will have to live with it at least until the fix is released in Cruise Control. I think it was merged yesterday, but not sure when it will be released. SnakeYAML does not offer backwards compatibility so I think just replacing it will not work.

**Reply by imrohankataria** [reply]
Great thanks! What is the default user for the Strimzi?

I see following, I dont want to select root and end the docker file without switching to the default use. Thanks

oot:x:0:0:root:/root:/bin/bash
bin:x:1:1:bin:/bin:/sbin/nologin
daemon:x:2:2:daemon:/sbin:/sbin/nologin
adm:x:3:4:adm:/var/adm:/sbin/nologin
lp:x:4:7:lp:/var/spool/lpd:/sbin/nologin
sync:x:5:0:sync:/sbin:/bin/sync
shutdown:x:6:0:shutdown:/sbin:/sbin/shutdown
halt:x:7:0:halt:/sbin:/sbin/halt
mail:x:8:12:mail:/var/spool/mail:/sbin/nologin
operator:x:11:0:operator:/root:/sbin/nologin
games:x:12:100:games:/usr/games:/sbin/nologin
ftp:x:14:50:FTP User:/var/ftp:/sbin/nologin
nobody:x:65534:65534:Kernel Overflow User:/:/sbin/nologin
dbus:x:81:81:System message bus:/:/sbin/nologin
systemd-coredump:x:999:997:systemd Core Dumper:/:/sbin/nologin
systemd-resolve:x:193:193:systemd Resolver:/:/sbin/nologin
kafka:x:1001:0::/home/kafka:/bin/bash

**Reply by imrohankataria** [reply]
```
FROM quay.io/strimzi/kafka:0.36.1-kafka-3.5.1

#user with permission to delete
USER root
# Remove the Kafka exporter and Cruise Control
RUN rm -rf /opt/kafka-exporter /opt/cruise-control && ls /opt

#switching back to custom user
USER kafka```

**Reply by scholzj** [reply]
Yeah, something like this should work. `kafka` is the user here.

**Reply by imrohankataria** [reply]
Strange, still coming up 
<img width="210" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/21262d6b-4a3a-4e26-aa0c-6afb5dd6395f">
<img width="1558" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/c230aab0-9c95-45f8-905c-e4ff377901e1">

**Reply by imrohankataria** [reply]
As per my investigation, its saved in the shasum file. Where are they usually saved? and can they be deleted?

**Reply by scholzj** [reply]
I guess you might need to squash the layers or something. But no idea how to do that I'm afraid. 🤷

**Reply by imrohankataria** [reply]
Strange errors happening now, Im able to see all the security vulnerabilities resolved but .sh commands not working

1. exec /opt/kafka/kafka_connect_run.sh: exec format error
2. 
```
INFO[0111] Initializing snapshotter ...                 
INFO[0111] Taking snapshot of full filesystem...        
INFO[0126] Cmd: /bin/sh                                 
INFO[0126] Args: [-c 'mkdir' '-p' '/opt/kafka/plugins/kafka-connect-jdbc/627a95aa'       && 'curl' '-f' '-L' '--output' '/opt/kafka/plugins/kafka-connect-jdbc/627a95aa/627a95aa.jar' 'https://packages.confluent.io/maven/io/confluent/kafka-connect-jdbc/10.7.0/kafka-connect-jdbc-10.7.0.jar'] 
INFO[0126] Util.Lookup returned: &{Uid:0 Gid:0 Username:root Name: HomeDir:/root} 
INFO[0126] Performing slow lookup of group ids for root 
INFO[0126] Running: [/bin/sh -c 'mkdir' '-p' '/opt/kafka/plugins/kafka-connect-jdbc/627a95aa'       && 'curl' '-f' '-L' '--output' '/opt/kafka/plugins/kafka-connect-jdbc/627a95aa/627a95aa.jar' 'https://packages.confluent.io/maven/io/confluent/kafka-connect-jdbc/10.7.0/kafka-connect-jdbc-10.7.0.jar'] 
error building image: error building stage: failed to execute command: starting command: fork/exec /bin/sh: exec format error```

**Reply by scholzj** [reply]
Wrong platform? Are you e.g. building it as Aarch64 image on Arm-based Mac and trying to run it on amd64 or something like that?

**Reply by imrohankataria** [reply]
you're right
on docker inspect <image> it comes **arm** and then AKS is **amd**

**Reply by scholzj** [reply]
Use this command to build amd64 images on my Arm-based mac: `docker build --platform linux/amd64 ...` ... not sure if it works everywhere.

**Reply by imrohankataria** [reply]
All works fine. I will close this out in a few if no other issue comes.

**Reply by imrohankataria** [reply]
How to go about these high issues within Strimzi? This is for com.google.guava, python-3libs, pip:9.0.3 ?
<img width="1196" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/41874450/ce158474-bd94-43ac-b044-790411232d9b">

**Reply by scholzj** [reply]
You would need to look into them, where they are and see if you can somehow fix them. You should also thing whether you are really affected by any of these.

**Comment by imrohankataria** [comment]
Wow, squash worked. Next step - test this image

**Reply by imrohankataria** [reply]
If anyone is following this thread, follow below steps -
1. cd ~/.docker/
2. VIM daemon.json
3. Change experimental to true 
4. docker build --squash -t img:tag .

**Reply by imrohankataria** [reply]
@scholzj looks like this feature is not available in docker anymore.
Do you have any guidelines on how to create a multi-stage docker image for strimzi? 

I have been trying something like below but doesn't work

```# Stage 1: Base Strimzi Kafka image
FROM quay.io/strimzi/kafka:0.36.1-kafka-3.5.1 as builder

# Switch to root user to make modifications
USER root

# Remove unwanted components
RUN rm -rf /opt/kafka-exporter /opt/cruise-control

# Remove the old guava library
RUN rm /opt/kafka/libs/guava-30.1-jre.jar

# # Stage 2: Final Alpine-based image
FROM alpine:latest

# # Create Kafka user with similar properties to Strimzi's Kafka user
RUN addgroup -g 1001 -S kafka && adduser -u 1001 -S kafka -G kafka

# # Copy necessary files from the builder stage, retaining permissions
COPY --from=builder --chown=kafka:kafka /opt/kafka /opt/kafka

# # Add full permissions for the 'kafka' group to the /opt/kafka directory
RUN chmod -R g+rwx /opt/kafka

# Switch to Kafka user
USER kafka```

**Reply by scholzj** [reply]
This is not really how it works. You cannot just change the base image like this. Especially not to Alpine that is typically missing many libraries available on other Linux distributions.

**Reply by imrohankataria** [reply]
@scholzj Can you please put me in a right direction? 
I need to get rid of other layers to be CVE free

**Reply by scholzj** [reply]
I don't know how to do it. I assume there will be other tools to squash the layers? I'm afraid you will have to figure it out yourself.
